core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala

```diff
@@ -622,67 +633,107 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
    *                      those failures be counted to task failure limits?
    * @param force whether to force kill busy executors, default false
+   * @param blacklistingOnTaskCompletion whether the executors are being killed due to
+   *                                     blacklisting triggered by the task completion event
    * @return the ids of the executors acknowledged by the cluster manager to be removed.
    */
   final override def killExecutors(
       executorIds: Seq[String],
       adjustTargetNumExecutors: Boolean,
       countFailures: Boolean,
-      force: Boolean): Seq[String] = {
+      force: Boolean,
+    blacklistingOnTaskCompletion: Boolean): Seq[String] = {
     logInfo(s"Requesting to kill executor(s) ${executorIds.mkString(", ")}")
 
-    val response = synchronized {
-      val (knownExecutors, unknownExecutors) = executorIds.partition(executorDataMap.contains)
-      unknownExecutors.foreach { id =>
-        logWarning(s"Executor to kill $id does not exist!")
-      }
-
-      // If an executor is already pending to be removed, do not kill it again (SPARK-9795)
-      // If this executor is busy, do not kill it unless we are told to force kill it (SPARK-9552)
-      val executorsToKill = knownExecutors
-        .filter { id => !executorsPendingToRemove.contains(id) }
-        .filter { id => force || !scheduler.isExecutorBusy(id) }
-      executorsToKill.foreach { id => executorsPendingToRemove(id) = !countFailures }
-
-      logInfo(s"Actual list of executor(s) to be killed is ${executorsToKill.mkString(", ")}")
-
-      // If we do not wish to replace the executors we kill, sync the target number of executors
-      // with the cluster manager to avoid allocating new ones. When computing the new target,
-      // take into account executors that are pending to be added or removed.
-      val adjustTotalExecutors =
-        if (adjustTargetNumExecutors) {
-          requestedTotalExecutors = math.max(requestedTotalExecutors - executorsToKill.size, 0)
-          if (requestedTotalExecutors !=
-              (numExistingExecutors + numPendingExecutors - executorsPendingToRemove.size)) {
-            logDebug(
-              s"""killExecutors($executorIds, $adjustTargetNumExecutors, $countFailures, $force):
-                 |Executor counts do not match:
-                 |requestedTotalExecutors  = $requestedTotalExecutors
-                 |numExistingExecutors     = $numExistingExecutors
-                 |numPendingExecutors      = $numPendingExecutors
-                 |executorsPendingToRemove = ${executorsPendingToRemove.size}""".stripMargin)
-          }
-          doRequestTotalExecutors(requestedTotalExecutors)
-        } else {
-          numPendingExecutors += executorsToKill.size
-          Future.successful(true)
+    var response: Future[Seq[String]] = null
+    val idleExecutorIds = executorIds.filter { id => !scheduler.isExecutorBusy(id) }
+    if (!blacklistingOnTaskCompletion) {
```

Comment:
What would be the performance cost to always use `makeOffersLock` (and deplementing the flag `blacklistingOnTaskCompletion`)? As this code is already quite complex and with the boolean flag dependent locking I think it will be even harder to follow.

Reply:
I agree that the code fix is a little tricky here, however, as far as I have tested, I have not seen a performance degradation in the job running time by addition of the extra lock.

Reply:
Sorry, there must be some misunderstanding here.
My suggestion is removing this if condition completely and using always:

```
makeOffersLock.synchronized {
  response = synchronized {
    killExecutorsImpl(idleExecutorIds, adjustTargetNumExecutors, countFailures, force)
  }
}
```
And as you got rid of the if you can remove the  `blacklistingOnTaskCompletion` from the methods's arguments as well.

As the order of locking always starts `makeOffersLock` I think this should be enough to avoid the deadlock.

Reply:
The flag `blacklistingOnTaskCompletion` is needed to ensure that the thread "task-result-getter-x" should not try to acquire the lock on `makeOffersLock` which is a necessary condition to avoid the deadlock between "task-result-getter" thread and "dispatcher-event-loop" thread. 

The reason is that when "task-result-getter" thread reaches the method `killExecutors()`, it has already acquired the lock on `TaskSchedulerImpl` and will try to acquire `makeOffersLock`. The "dispatcher-event-loop" thread on the other hand, acquires `makeOffersLock` and will wait on acquiring `TaskSchedulerImpl` lock in the method `resourceOffers()`, thus leading to the deadlock.

Reply:
Ok I see. I checked the first deadlock and I think the problem is in `org.apache.spark.scheduler.TaskSchedulerImpl#isExecutorBusy`:

https://github.com/apache/spark/blob/b15423361bc28c4cd2216683eb852fdbec3ea58f/core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala#L824-L826
   
That `synchronised` is too restrictive here for reading a snapshot state of the `executorIdToRunningTaskIds` map. For this problem a solution could be just using [TrieMap](https://www.scala-lang.org/api/current/scala/collection/concurrent/TrieMap.html), which is "A concurrent hash-trie or TrieMap is a concurrent thread-safe lock-free implementation of a hash array mapped trie". 

If you change the type of `executorIdToRunningTaskIds` from HashMap to TrieMap then you can remove the synchronised from `isExecutorBusy`. 

I have checked and the `isExecutorBusy` is only used from two places:
- org.apache.spark.scheduler.TaskSchedulerImpl#resourceOffers where we already in a synchronised block, so with the type change the behaviour is the same as before
-  org.apache.spark.scheduler.cluster.CoarseGrainedSchedulerBackend#killExecutors where we already lived with a snapshot state which could be outdated after the method call

Regarding the second deadlock I will continue my analyses.

Reply:
We could but, the values of the `executorIdToRunningTaskIds` map (`HashSet`) are not protected or thread safe. We leave the protection of the concurrent trie map after the `.get(execId)`, then there are race conditions on the `.exists` that happens against this hash set. We could make the `HashSet` a concurrent trie also.

This could result in `isExecutorBusy` telling you an executor was idle when it wasn't, or the other way around. Leading to some of the same things that were fixed in https://issues.apache.org/jira/browse/SPARK-19757 coming back.

Reply:
@attilapiros Believe me, making `executorIdToRunningTaskIds` a TrieMap did come to my mind and I did give it a shot. It works for our purposes, but we realized something else. The above map is of type <String, HashSet>. This map might protect the reference to the HashSet object, however, it gives no guarantee that the HashSet will not be mutated while accessing the map. It basically does not guarantee protection on the HashSet. It would have been much easier to implement the TrieMap had the value of the map would have simply been an Object and not a collection. I could be wrong here, so if you feel otherwise, some pointers would be helpful. Thank you.

Reply:
@pgandhi999  using `collection.immutable.Set` as the value type in the TrieMap would solve that.

Reply:
I see there is https://issues.apache.org/jira/browse/SPARK-19757 and my solution does not solves that. I will think about that one too.

Reply:
For the second deadlock and SPARK-19757 I would in the thread "spark-dynamic-executor-allocation" lock in the order of the other "task-result-getter-0", scheduler first:

```scala
    // Make fake resource offers on all executors
    private def makeOffers() {
      // Make sure no executor is killed while some task is launching on it
      val taskDescs = scheduler.synchronized {        // <-------- extra lock
        CoarseGrainedSchedulerBackend.this.synchronized {
          // Filter out executors under killing
          val activeExecutors = executorDataMap.filterKeys(executorIsAlive)
          val workOffers = activeExecutors.map {
            case (id, executorData) =>
              new WorkerOffer(id, executorData.executorHost, executorData.freeCores,
                Some(executorData.executorAddress.hostPort))
          }.toIndexedSeq
          scheduler.resourceOffers(workOffers)
        }
      }
      if (!taskDescs.isEmpty) {
        launchTasks(taskDescs)
      }
    }

```

Also in the org.apache.spark.scheduler.cluster.CoarseGrainedSchedulerBackend.DriverEndpoint#makeOffers.

TrieMap with immutable collection.immutable.Set and using the same order of locking would solve the deadlocks and SPARK-19757 too I think.

Reply:
Without having read the code or the bugs, just remember that a trie, IIRC, has `O(log n)` insert and retrieval time.

Reply:
Yes, I just focused on saving the extra lock first. 

But we could keep track of the executor IDs where tasks are scheduled/running separately in a concurrently accessable set (volatile reference for an Immutable Set or CopyOnWriteArraySet). 

The method `isExecutorBusy` could use this new set. So we can keep HashMap for `executorIdToRunningTaskIds` and still we are not introducing that lock.

Reply:
I don't think that `makeOffersLock` solves the deadlock here.  You wont' get a deadlock between the same two locks, but now it can be with `makeOffersLock` instead.  Consider this sequence (some simplification of full call stack, but showing the important locks at least)

1) taskresultgetter: handleFailedTask --> lock on taskSchedulerImpl
2) taskresultgetter: BlacklistTracker.killExecutor

3) dispatcher: receive --> lock on CoarseGrainedSchedulerBackendkk
4) dispatcher: makeOffers --> lock on makeOffersLock
5) dispatcher: blocked on TaskSchedulerImpl lock

6) taskResultGetter: makeOffers, but blocked on makeOffersLock

As Attila suggested, I would consider creating an ordering between the TaskSchedulerImpl lock and the CoarseGrainedSchedulerBackend lock, so that we always get the TaskSchedulerImpl lock first.  Of course that comes with a performance penalty, and we will have to audit all other uses of the CoarseGrainedSchedulerBackend lock too.

Still thinking about any other options ...

Reply:
@squito I agree with you and @attilapiros about creating an ordering. I shall definitely follow the approach and try it out. 

Regarding your comment on the deadlock between `makeOffersLock` and task-result-getter thread, that should ideally not happen as the task-result-getter thread will never compete for acquiring `makeOffersLock`. The reason I have added the flag `blacklistingForTaskCompletion` is to ensure that task-result-getter thread never acquires lock on `makeOffersLock`.

Also, you are right in saying that `makeOffersLock` does not solve the deadlock. I have explained the purpose of `makeOffersLock` in my comment below. Quoting it here:

> I can explain more about the makeOffersLock here.
> 
> PR #17091 introduced the part about acquiring synchronization on CoarseGrainedSchedulerBackend object in the method makeOffers(). This particular piece of code introduced a deadlock between task-result-getter thread and dispatcher-event-loop thread. I can simply removed the synchronized statement in makeOffers() and the deadlock would be resolved and we really do not need makeOffersLock.
> 
> However, removing the synchronized statement will once again expose the race condition described in JIRA https://issues.apache.org/jira/browse/SPARK-19757 for which the fix in the corresponding PR was merged. makeOffersLock here serves as the solution to the above problem. By synchronizing on makeOffersLock, the race condition between dynamic-executor-allocation thread and dispatcher-event-loop thread is avoided. That is indeed it's sole purpose. I am however, open to discussing and working on better solutions to the above problem, if any. Thank you.

Reply:
I am basically trying to solve the two deadlocks and also fix the race condition issue for SPARK-19757. I think the approach of ordering alongwith using a ConcurrentHashMap with immutable Set as suggested by @attilapiros and @squito should work out. Let me work on that and get back to you.

Reply:
ah I see, sorry I misread part of the logic, thanks